src/utilities/urlSplitter.ts

```diff
@@ -0,0 +1,30 @@
+type SplitURL = {
+  protocol: string | undefined,
+  domain: string | undefined,
+  port: string | undefined,
+  path: string | undefined,
+  query: string | undefined,
+  hash: string | undefined,
+}
+
+export function splitUrl(url: string): SplitURL {
+  const urlPattern = /^(?:(?<protocol>\w+):\/\/)?(?:([^/@:]+)(?::([^/@:]+))?@)?(?<domain>[^/?#:]+)?(?:(?<port>:\d+))?(?<path>\/[^?#]*)?(?<query>\?[^#]*)?(?<hash>#.*)?$/
```

Comment:
this was written by GPT, could probably be cleaned up

Reply:
Couldn't we just use the URL class for this?

Reply:
It's very possible, but my first attempt resulted in constant errors being thrown for "invalid urls". I'll do a little research and report back

Reply:
Most likely do to a lack of complete urls. That class needs a fully qualified url. But you can also provide it a base to use.